autoblocks/_impl/tracer.py

```diff
@@ -103,44 +134,119 @@ def start_span(self) -> Generator[None, None, None]:
                 props["span_id"] = prev_span_id
             self.set_properties(props)
 
+    def _evaluation_to_json(self, evaluation: Evaluation, evaluator_external_id: str) -> Dict[str, Any]:
+        return dict(
+            id=str(uuid.uuid4()),
+            score=evaluation.score,
+            threshold=dataclasses.asdict(evaluation.threshold) if evaluation.threshold else None,
+            metadata=evaluation.metadata,
+            evaluatorExternalId=evaluator_external_id,
+        )
+
+    async def _evaluate_event(self, event: TracerEvent, evaluator: BaseEventEvaluator) -> Optional[Evaluation]:
+        """
+        Evaluates an event using a provided evaluator.
+        """
+        evaluation = None
+        if inspect.iscoroutinefunction(evaluator.evaluate_event):
+            try:
+                evaluation = await evaluator.evaluate_event(event=event)
+            except Exception as err:
+                log.error(f"Unable to execute evaluator with id: {evaluator.id}. Error: %s", err, exc_info=True)
+        else:
+            try:
+                ctx = contextvars.copy_context()
+                evaluation = await global_state.event_loop().run_in_executor(
+                    None,
+                    ctx.run,
+                    evaluator.evaluate_event,
+                    event,
+                )
+            except Exception as err:
+                log.error(f"Unable to execute evaluator with id: {evaluator.id}. Error: %s", err, exc_info=True)
+
+        return evaluation
+
+    async def _run_evaluators_unsafe(
+        self, evaluators: List[BaseEventEvaluator], event: TracerEvent, max_evaluator_concurrency: int
+    ) -> List[Dict[str, Any]]:
+        if len(evaluators) == 0:
+            return []
+        evaluations: List[Evaluation] = await gather_with_max_concurrency(
+            max_evaluator_concurrency,
+            [
+                self._evaluate_event(
+                    event=event,
+                    evaluator=evaluator,
+                )
+                for evaluator in evaluators
+            ],
+        )
+        return [
+            self._evaluation_to_json(evaluation=evaluation, evaluator_external_id=evaluator.id)
+            for evaluator, evaluation in zip(evaluators, evaluations)
+            if evaluation is not None
+        ]
+
     async def _send_event_unsafe(
         self,
         # Require all arguments to be specified via key=value
         *,
         message: str,
-        trace_id: Optional[str] = None,
-        span_id: Optional[str] = None,
-        parent_span_id: Optional[str] = None,
-        timestamp: Optional[str] = None,
-        properties: Optional[Dict[Any, Any]] = None,
-        prompt_tracking: Optional[Dict[str, Any]] = None,
-    ) -> SendEventResponse:
+        max_evaluator_concurrency: int,
```

Comment:
can we move this arg to the bottom next to `evaluators` like in the other signature

Reply:
Updated